tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/GraphQL/GraphQLSecurityCommon.cs

```diff
@@ -0,0 +1,116 @@
+// <copyright file="GraphQLSecurityCommon.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+using System.Web;
+using Datadog.Trace.AppSec;
+using Datadog.Trace.AppSec.Coordinator;
+
+namespace Datadog.Trace.ClrProfiler.AutoInstrumentation.GraphQL;
+
+internal sealed class GraphQLSecurityCommon
+{
+    private static GraphQLSecurityCommon _instance;
+
+    private readonly Dictionary<IScope, Dictionary<string, object>> _scopeResolvers;
+
+    private GraphQLSecurityCommon()
+    {
+        _scopeResolvers = new();
+    }
+
+    private static GraphQLSecurityCommon GetInstance()
+    {
+        return _instance ??= new GraphQLSecurityCommon();
+    }
+
+    private Dictionary<string, object> GetScopeResolvers(IScope scope)
+    {
+        if (!_scopeResolvers.TryGetValue(scope, out var resolvers))
+        {
+            resolvers = new();
+            _scopeResolvers.Add(scope, resolvers);
+        }
+
+        return resolvers;
+    }
+
+    private void RemoveScopeResolvers(IScope scope)
+    {
+        _scopeResolvers.Remove(scope);
+    }
+
+    internal static void RegisterResolverCall(IScope scope, string resolverName, Dictionary<string, object> arguments)
+    {
+        var resolvers = GetInstance().GetScopeResolvers(scope);
+
+        if (!resolvers.TryGetValue(resolverName, out var resolverCalls))
+        {
+            resolverCalls = new List<object>();
+        }
+
+        try
+        {
+            resolvers.Add(resolverName, resolverCalls);
+        }
+        catch (ArgumentException)
```

Comment:
Why would you get an argument exception, when the resolver is already present? It would better to only add if it's not present, which could be done as part of the previous if case.

If you're worried about concurrency issues, i.e. failing because the dictionary's already been changed by another thread then a concurrent dictionary would probably be a better choice of structure.

Reply:
Yea I miss it, just adding it in the previous if case will do it.

I think there is only one thread by request (and one root dictionary for a given scope), so no concurrency issues should occur here, but if that's a better implementation without reducing performance, I can change to this type of dictionary.